## <font color=#0099CC>mIAx - Taller Renta Fija - ANÁLISIS CARTERA DE RENTA FIJA</font>

En esta práctica, desarrollaremos un análisis relativamente exahustivo de un universo de Renta Fija, en concreto, de bonos corporativos. Además, construiremos y analizaremos varias carteras. 

Para ello, contaremos con la siguiente información almacenada en la carpeta *data*:
- Universo de bonos, con sus características esenciales (fichero *universo.csv*)
- Histórico de precios de cierre del universo de bonos anterior (fichero *precios_historicos_universo.csv*)
- Curva de tipos de interés €STR (fichero *curvaESTR.csv*)
- Histórico de precios de otros índices que nos serán de utilidad (fichero *precios_historicos_varios*):
    - Índices de crédito: ITRAXX Main y ITRAXX XOVER. Serán útiles para la cobertura del riesgo de crédito.
    - Futuros sobre el *Schatz* (DU1), *BOBL* (OE1) y *BUND* (RX1). Serán útiles para la cobertura de los tipos de interés.
    - Índice de crédito *RECMTREU*, que valdría como benchmark de las carteras que construyamos.

No necesariamente se usará toda toda la información

En líneas generales, estos son los ejercicios que completaremos, aunque los detallaremos más en cada apartado:
1. Análisis de datos. En esta sección, haremos un análisis de la información que tenemos de cada bono y lo que significa. Asimismo, haremos los tratamientos y limpieza que necesitemos para luego poder usarlos.
2. Valoración de los bonos del universo utilizando la curva de descuento y bajo ciertas asunciones. Comparación de estos precios con los precios de mercado.
3. Cálculo del spread que pagan los bonos sobre la curva.
4. Cálculo de *yield*, duración y convexidad.
5. Contrucción de una cartera equiponderada con todos los bonos del universo. Contraste con el benchmark (os proponemos el índice RECMTREU para el que os hemos dado los precios) y backtest de la estrategia. ¡OJO! El índice es *Total Return*.
6. Tienes el mandato de construir una cartera de como máximo **20** bonos corporativos con ese universo y una serie de restricciones y, claro, maximizando la rentabilidad total de la cartera:
    - La duración de la cartera no debe superar los 3 años
    - La exposición a emisiones HY no puede superar el 10% de la cartera
    - No puedes invertir en deuda subordinada
    - No se puede invertir en emisiones de tamaño igual o inferior a 500 millones
    - No se puede invertir más de un 10% del capital en una misma emisión
    - No puede haber más de un 15% de concentración en un mismo emisor
    (¡OJO! No estamos teniendo en cuenta en este ejercicio si hubiera un mínimo de inversión, lo cuál sería un dato relevante tener en cuenta en un caso real)

    6.1. Constrúyela a fecha de hoy

    6.2. Teniendo en cuenta la naturaleza que nos están pidiendo para la cartera, ¿añadirías alguna otra restricción?

    6.3. ¿Cómo medirías el riesgo de crédito de la cartera?

    6.4. ¿Cómo medirías el riesgo de liquidez de la cartera?

    6.5. Describe cómo habría que hacer el backtest de esta cartera
7. Ahora, se te pide que cubras la exposición de la cartera a los tipos de interés. Con la información que tienes, ¿cómo lo harías?
8. ¿Y si quisieras cubrir total o parcialmente el riesgo de crédito? Usa de nuevo la información que tienes.
9. ¿Cómo construirías tu cartera? ¿Se te ocurre alguna estrategia específica, por ejemplo, de valor relativo?


📣 <font color=#CC6600>**¡NORMAS!**</font>

La práctica se puede hacer en grupos de hasta **3 personas** y deberá entregarse antes del **27 de noviembre**. 

Cada grupo expondrá una parte de los ejercicios en la clase del día 27, donde la resolveremos juntos a modo de taller. Esta exposición contará hasta **1 punto** de la nota final.

Además, se valorará positivamente para la práctica la participación en las clases.

📣 <font color=#CC6600>**¡IMPORTANTE!**</font>

Todo el código implementado debe estar debidamente comentado e incluir conclusiones de los resultados obtenidos para optar a la máxima puntuación. Asimismo, se debe responder a las preguntas planteadas.

Las propuestas de mejora o posibles trabajos futuros se valorarán positivamente también.

Usa las fuentes de información que consideres necesarias para apoyar tus respuestas.

### <font color=#336699>Librerías</font>

In [ ]:
import matplotlib.pyplot as plt
import pandas as pd
import datetime

In [2]:
# Estilo
plt.style.use('dark_background')

In [3]:
# Fecha de análisis
fecha_analisis = pd.to_datetime('2025-10-01')

### <font color=#336699>1. Datos</font>

<style>.gray {background-color: #595959}

</style><div class="gray">❕💬¿Qué observas en los datos? Analiza la información que tenemos del universo</div>

Haz un primer análisis visual de la información que tenemos del universo de bono.
Entre otras, plantéate cuestiones como:
- ¿Divisas?
- ¿Tipo de bonos? ¿Fijo/Flotante? ¿Prelación? ¿Opcionalidad? ¿Hay bonos perpetuos?
- ¿Sectores? ¿Emisores? Si invirtéramos en todos los bonos, ¿dirías a priori que la cartera está diversificada?
- ¿Ratings? (Riesgo de crédito)
- ¿Otros datos cuantitativos?
    - Riesgo de liquidez - Horquillas y nominal vivo
- ¿Hay *gaps* en la información que vamos a tener que tratar?

In [ ]:
# Cargar y analizar el universo de bonos
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Cargar datos
universo = pd.read_csv('../data/universo.csv', sep=';', parse_dates=['Maturity', 'Next Call Date'], dayfirst=True)
print(f"📊 ANÁLISIS DEL UNIVERSO DE BONOS")
print(f"Total bonos: {len(universo)}")

# Análisis por divisas
print(f"\n💰 DIVISAS:")
print(universo['Ccy'].value_counts())

# Análisis por tipo de bono
print(f"\n🏦 TIPOS DE BONO:")
print(f"Fijos: {(universo['Coupon Type'] == 'FIXED').sum()}")
print(f"Flotantes: {(universo['Coupon Type'] == 'FLOATING').sum()}")
print(f"Perpetuos: {universo['Maturity'].isna().sum()}")
print(f"Callable: {(universo['Callable'] == 'Y').sum()}")

# Análisis por seniority
print(f"\n⚖️ SENIORITY:")
print(universo['Seniority'].value_counts())

print(f"\n✅ Universo completamente en EUR, diverso en estructura y seniority")

#### <font color=#808080>Divisas</font>

In [ ]:
La única divisa presente en el universo es 'EUR'

#### <font color=#808080>Tipos de bono</font>

In [ ]:
Todos los bonos son corporativos.
Podemos ver que hay 1910 bonos fijos y 345 flotantes. Lo que supone un 84.70% de los bonos son fijos y el 15.3% restantes son flotantes.
Ademas solo 19 bonos son perpetuos. Lo que supone un 0.84% de los bonos son perpetuos.
En cuanto a la opcionalidad; 1640 bonos son callable y los 615 bonos restantes no lo son.
Por último en cuanto a la prelacion podemos distinguir los distintos tipos:
1st lien              13
Jr Subordinated       25
Secured               20
Sr Non Preferred     190
Sr Preferred         179
Sr Unsecured        1675
Subordinated         153


#### <font color=#808080>Sectores y emisores</font>

In [ ]:
Podemos distinguir varios sectores que emiten los bonos, los cuales son los siguientes:
Basic Materials             72 bonos
Communications             175 bonos
Consumer, Cyclical         171 bonos
Consumer, Non-cyclical     400 bonos
Energy                      13 bonos
Financial                 1012 bonos
Industrial                 211 bonos
Technology                  51 bonos
Utilities                  150 bonos

En cuanto a los emisores podemos destacar los 10 emisores que más bonos han emitido, los cuales son los siguientes:
BANQUE FED CRED MUTUEL     33 bonos 
BNP PARIBAS                27 bonos
VONOVIA SE                 23 bonos
ORANGE SA                  23 bonos
BPCE SA                    23 bonos
SOCIETE GENERALE           21 bonos
VOLKSWAGEN INTL FIN NV     21 bonos
NESTLE FINANCE INTL LTD    20 bonos
UNIBAIL-RODAMCO-WESTFLD    19 bonos
CREDIT AGRICOLE SA         19 bonos

#### <font color=#808080>Ratings (Riesgo de crédito)</font>

In [ ]:
En cuanto al rating podemos observar como varia desde BB+ hasta AAA. Pero el rating que más se repite es el de BBB+, con un total de 439 bonos 
y el que menos AAA con solo 4 bonos

#### <font color=#808080>Riesgo de liquidez - Horquillas y nominal vivo</font>

In [ ]:
En cuanto al nominal vivo, tras un analisis, podemos observar que el valor que más se repiten vas desde 440.000€ hasta 5200.000€. Esta cantidad se repite en
686 bonos. Por lo que entendemos que alrededor de 500.000€ de 686 bonos emitidos no se han comprado. 

#### <font color=#808080>Resto de información</font>

En esta sección, analizamos el resto de ficheros para ver qué información tenemos y, en caso de haber *gaps*, limpiar los datos antes de trabajar con ellos.

##### <font color=#CC6600>Precios bonos universo</font>

In [ ]:
En el fichero precios_historicos_universo.csv podemos ver una serie historica de precios del bonos, desde 01/10/2023 hasta el 01/10/2025


##### <font color=#CC6600>Otros precios</font>

Para terminar con el análisis de datos, falta lo leído en los ficheros de *"precios_historicos_varios.csv"* y *curvaESTR.csv*.

ANÁLISIS DE PRECIOS HISTÓRICOS VARIOS
Dimensiones: (514, 6)
Período: 2023-10-02 00:00:00 a 2025-10-01 00:00:00

Instrumentos disponibles:
  ITRX EUR CDSI GEN 5Y Corp: 506 observaciones
  ITRX XOVER CDSI GEN 5Y Corp: 506 observaciones
  DU1 Comdty: 509 observaciones
  OE1 Comdty: 509 observaciones
  RX1 Comdty: 509 observaciones
  RECMTREU Index: 514 observaciones

📈 INSTRUMENTOS CLAVE:
• ITRAXX Main: Índice CDS grado inversión (IG)
• ITRAXX XOVER: Índice CDS alto rendimiento (HY)
• DU1/OE1/RX1: Futuros alemanes (Schatz/BOBL/BUND)
• RECMTREU: Benchmark índice Total Return

ANÁLISIS CURVA €STR
Fecha de referencia: 2025-10-01 00:00:00
Puntos de la curva: 33
Vencimiento más largo: 2075-10-03 00:00:00

Puntos clave de la curva:
  0.0Y: Market=1.927%, Zero=1.954%, DF=0.999518
  0.3Y: Market=1.927%, Zero=1.949%, DF=0.994887
  0.7Y: Market=1.907%, Zero=1.921%, DF=0.987187
  1.5Y: Market=1.889%, Zero=1.901%, DF=0.971717
  6.0Y: Market=2.240%, Zero=2.329%, DF=0.869409
  50.0Y: Market=2.768%, Zero=2.758%, DF=0.251512

### <font color=#336699>2. Valoración</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 En esta sección, valoraremos los bonos utilizando la curva. Para ello, crea una función (puedes hacerlo en un .py aparte) que con las **características del bono, la curva y un spread de crédito** devuelva la valoración del bono (incluyendo **precio limpio, cupón corrido y precio sucio**).

Si asumimos que el **spread de crédito es 0**, y la ejecutamos para el 01/10/2025...
- ¿Qué observas si comparas los precios obtenidos y los precios de mercado? 
- ¿Crees que la diferencia se debe a un factor relacionado sólo con el riesgo crediticio?
- ¿Qué otros factores influyen en ese spread?

Para la valoración, haz las siguientes simplificaciones:

- Asume que el vencimiento de los bonos perpetuos (para los que no hay vencimiento) es la próxima fecha call.
- Asume que todos aquellos bonos que tengan call serán calleados. Por lo tanto, usa la fecha call como fecha de vencimiento.
- Asume que los cupones son fijos hasta vencimiento (aunque alguno cambie a lo largo de la vida del bono).
- Usa la base de cálculo ACT/365. No tengas en cuenta la convención de día hábil.

Ten en cuenta que necesitarás una función de interpolación también. Interpola los factores de descuento exponencialmente.

</div>

In [ ]:
¿Qué observas si comparas los precios obtenidos y los precios de mercado?
Asumiendo un spread de crédito del 0% observamos que los precios calculados son significativamente más altos que los precios de mercado. Esto significa por ejemplo que si el bono cotiza a 104, el modelo devuelve 108.
La diferencia es positiva (Precio Calculado > Precio Mercado). Esto significa que el mercado está descontando los flujos a una tasa mucho más alta que la curva ESTR (libre de riesgo).

¿Crees que la diferencia se debe a un factor relacionado sólo con el riesgo crediticio?
No, no es solo riesgo de crédito. Aunque es el componente principal (dado que son bonos corporativos con ratings tipo BBB)
Asunción de Call (La gran distorsión): Todos los bonos con opción de Call se amortizan en la próxima fecha de call.
El problema: Un emisor solo ejercerá la Call si le beneficia (generalmente si puede refinanciarse más barato). Si un bono en el mercado cotiza con descuento (ej. precio 90) o bajo la par, es muy probable que el emisor no lo calleé, porque no va a pagar 100 por una deuda que el mercado valora a 90.
Al forzar el vencimiento a la fecha de Call, estás valorando como si fueras a recibir el principal (100%) muy pronto, lo que dispara artificialmente el precio calculado frente al de mercado, el cual asume que el bono vivirá hasta su vencimiento real porque no es óptimo callearlo.

¿Qué otros factores influyen en ese spread?
La diferencia entre el precio (Riesgo Cero + Call Forzosa) y el precio de mercado se compone de:
Prima de Riesgo de Crédito (Credit Spread): La compensación por la posibilidad de default del emisor (JDE Peet's, Deutsche Telekom, etc.).
Prima de Liquidez: Los bonos corporativos son menos líquidos que los bonos del gobierno o los swaps ESTR; los inversores exigen rentabilidad extra por ello.
Valor de la Opcionalidad (Convexidad negativa): Como mencionamos arriba, el mercado valora la probabilidad de que el bono sea calleado o no. El modelo asume probabilidad del 100%. Esa diferencia de opcionalidad es enorme.
Discrepancia de Curvas: Estamos usando ESTR (Overnight) para descontar. Muchos bonos corporativos se emiten referenciados a Euribor 6M o 3M. Existe una base (spread) entre ESTR y Euribor que el modelo no está capturando.

### <font color=#336699>3. Spread</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Calculemos ahora los spreads que debemos añadir a la curva con un movimiento paralelo para que cuadren los precios de mercado que tenemos. Para ello, usa la función de valoración del apartado anterior. 
- ¿Qué observas? ¿Tienen sentido los resultados? 
- ¿Con qué datos de los que tenemos compararías para ver si los resultados son coherentes?

</div>

In [ ]:
¿Que observas?¿Tienen sentido los resultados?
Observamos una distribución de spreads implicados que varía por emisor, vencimiento y seniority.
Los spreads tienden a incrementarse para emisiones de menor calidad crediticia y para emisiones con menor tamaño de outstanding.
Cabe destacar que los spread que obtenemos los tenemos en bps. 

¿Con qué datos de los que tenemos compararías para ver si los resultados son coherentes?
Compararía los spreads implícitos con: rating del emisor, PD 1YR, spreads de bonos comparables del mismo sector y seniority, tamaño de la emisión (outstanding), y bid-ask. 
También comprobaría si el precio de mercado es clean o dirty, la fecha de valoración y posibles eventos corporativos (callable, amortizaciones).

### <font color=#336699>4. YTM, Duración, Convexidad</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Calculemos ahora la siguiente información, también relacionada con la rentabilidad y riesgo de las emisiones:
- *Yield* - Por simplicidad, en este caso, en el caso de los bonos callable, nos quedaremos con la fecha call, como en el ejercicio anterior. Usa las mismas asunciones que para la valoración y el spread.
- Duración
- Convexidad

Responde a las siguientes preguntas:
- ¿Que relación hay entre la TIR calculada y el spread calculado en el apartado anterior?
- ¿Qué relación hay entre la duración y el vencimiento? ¿Qué refleja la duración? ¿De qué otra forma se podría obtener esta sensibilidad?
- Estima el precio del bono usando la duración y convexidad, ¿qué observas?

</div>

In [ ]:
¿Que ralacion hay entre la TIR calculada y el spread calculado en el apartado anterior?
La TIR (yield) es la tasa interna que iguala el precio con los flujos del bono. El spread calculado en el apartado anterior es el diferencial que hay
que añadir a la curva risk-free para que, descontando con esa curva ajustada, el precio del bono cuadre con el mercado. 
Ambos están directamente relacionados: un mayor spread implicado suele traducirse en una mayor yield para ese bono (todas las demás cosas iguales). 
Sin embargo, la relación no es uno a uno porque el spread se aplica sobre la curva risk-free y modifica directamente los factores de descuento de 
todos los flujos, mientras que la yield es una sola tasa que describe la rentabilidad agregada del bono. En la práctica, bonos con mayor spread 
implícito tendrán yields más altos; las diferencias entre spread y yield reflejan también la forma temporal de la curva risk-free y la estructura de 
cupones del bono.

¿Qué relación hay entre la duración y el vencimiento? ¿Qué refleja la duración? ¿De qué otra forma se podría obtener esta sensibilidad?
La duración mide el tiempo medio ponderado hasta la recepción de los flujos y la sensibilidad del precio ante cambios de yield. En general, para 
bonos cupón fijo y sin opción, la duración tiende a aumentar con el vencimiento: cupones largos retrasan el cobro del principal y por tanto aumentan
la media ponderada. Sin embargo, la relación no es lineal: para bonos con cupones, la duración crece con el vencimiento pero se aproxima a un 
límite (el bono perpetuo tiene una duración finita si tiene cupones). La duración refleja la sensibilidad (primera derivada) del precio frente a 
cambios en la yield.

Estima el precio del bono usando la duracion y convexidad, ¿que observas?
Pestimado = P0(1-ModDurAy + 1/2*Conv(Ay)^2)
Para cambios pequeños de yield (p. ej. hasta unos pocos decenas de puntos básicos) la estimación suele ser bastante buena: duración da la 
mayor parte del ajuste y la convexidad corrige la curvatura. Para cambios grandes en yield la aproximación pierde precisión y puede ser inexacta
porque la relación precio–yield es no lineal y las fórmulas de Taylor truncadas dejan de ser adecuadas. Además, si el bono tiene opciones (callables),
la sensibilidad efectiva puede diferir mucho (la duración efectiva cambia con el nivel de tipos), por lo que la aproximación puede fallar si la 
opción se vuelve in-/out-of-the-money tras el shock de yield.


### <font color=#336699>5. Cartera equiponderada</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Crea un algoritmo de inversión que consista en una cartera equiponderada, asignando el mismo peso a todos los bonos vivos en cada fecha de rebalanceo.

Asume rebalanceo mensual, y que no tenemos costes más allá de los implícitos en el propio precio, calcula la evolución que hubiese tenido tu algoritmo. Por simplificación, utiliza los precios MID que se te dan.

Asumiendo que el benchmark de la cartera es el índice que se nos da: *RECMTREU Index*. Contrasta la evolución de tú cartera contra dicho benchmark. Ten cuidado porque es un índice *Total Return*.

- ¿Qué sería lo más correcto en lugar de utilizar los precios MID?
- ¿Se te ocurre algún otro benchmark que se podría utilizar?

</div>

In [ ]:
¿Que seria lo más correcto en lugar de utilizar los precios MID?

¿Se te ocurre algun otro benchmark que se podría utilizar?

### <font color=#336699>6. Cartera mandato</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Como adelantábamos en el enunciado, tienes el mandato de construir una cartera de como máximo **20** bonos corporativos con ese universo y una serie de restricciones y, claro, maximizando la rentabilidad total de la cartera:
- La duración de la cartera no debe superar los 3 años
- La exposición a emisiones HY no puede superar el 10% de la cartera
- No puedes invertir en deuda subordinada
- No se puede invertir en emisiones de tamaño igual o inferior a 500 millones
- No se puede invertir más de un 10% del capital en una misma emisión
- No puede haber más de un 15% de concentración en un mismo emisor
(¡OJO! No estamos teniendo en cuenta en este ejercicio si hubiera un mínimo de inversión, lo cuál sería un dato relevante tener en cuenta en un caso real)

1. Teniendo en cuenta la naturaleza que nos están pidiendo para la cartera, ¿añadirías alguna otra restricción?

2. ¿Cómo medirías el riesgo de crédito de la cartera?

3. ¿Cómo medirías el riesgo de liquidez de la cartera? ¿Se te ocurre alguna otra información que se podría utilizar aunque no se te haya dado?

4. Describe cómo habría que hacer el backtest de esta cartera, no hace falta que lo implementes en este caso

</div>

In [ ]:
## 📋 **RESTRICCIONES ADICIONALES QUE AÑADIRÍA:**

### 1. **Diversificación Sectorial y Geográfica:**
- **Máximo 25% por sector** (Financiero, Utilities, Industrial, etc.)
- **Máximo 30% por país** (España, Francia, Alemania, Italia)
- **Razón**: Evitar concentraciones implícitas que podrían no ser evidentes a nivel de emisor individual

### 2. **Calidad Crediticia Promedio:**
- **Rating promedio ponderado ≥ BBB** 
- **Máximo 5% en bonos por debajo de BBB-**
- **Razón**: Mantener un perfil de riesgo conservador y predecible

### 3. **Restricciones de Liquidez:**
- **Mínimo volumen negociado diario ≥ 1M EUR**
- **Bid-Ask spread máximo ≤ 50 bps**
- **Razón**: Garantizar que podemos entrar/salir de posiciones sin impacto significativo

### 4. **Restricciones Temporales:**
- **Vencimiento mínimo > 1 año** (evitar riesgo de reinversión)
- **Máximo 40% con vencimiento < 2 años**
- **Razón**: Balance entre liquidez y estabilidad de ingresos

### 5. **Restricciones de Correlación:**
- **Máximo 20% en emisores del mismo grupo empresarial**
- **Límite en bonos con correlación > 0.8**
- **Razón**: Evitar riesgos sistémicos ocultos

##### <font color=#CC6600>Riesgo de crédito</font>

In [ ]:
## **MEDICIÓN DEL RIESGO DE CRÉDITO:**

### **Método 1: Spread Crediticio Ponderado**
```python
credit_risk = Σ(weight_i × credit_spread_i)
```
- **Ventaja**: Directamente observable en mercado
- **Limitación**: Puede ser volátil y afectado por liquidez

### **Método 2: Probabilidad de Default Ponderada (PD)**
```python
portfolio_PD = Σ(weight_i × PD_1Y_i)
```
- **Datos disponibles**: Columna 'PD 1YR' en nuestro universo
- **Interpretación**: Probabilidad esperada de default en 1 año

### **Método 3: Rating Score Ponderado**
```python
# Convertir ratings a scores numéricos
rating_scores = {'AAA': 1, 'AA+': 2, ..., 'D': 21}
portfolio_rating = Σ(weight_i × rating_score_i)
```

### **Método 4: Credit VaR (Más Sofisticado)**
- **Modelo**: Monte Carlo con correlaciones entre emisores
- **Output**: Pérdida esperada y inesperada por crédito
- **Datos necesarios**: Matriz de correlaciones, LGD, EAD

### **IMPLEMENTACIÓN EN NUESTRO CASO:**
```python
# Usando datos disponibles
credit_risk_spread = (portfolio['weight'] * credit_spreads).sum()
credit_risk_PD = (portfolio['weight'] * portfolio['PD_1YR']).sum()
```

##### <font color=#CC6600>Riesgo de liquidez</font>

In [ ]:
##  **MEDICIÓN DEL RIESGO DE LIQUIDEZ:**

### **Métrica 1: Disponibilidad de Precios**
```python
liquidity_score = (días_con_precio / días_totales) * 100
```
- **Interpretación**: % de días con cotización disponible
- **Umbral**: >90% para considerar líquido

### **Métrica 2: Bid-Ask Spread Promedio**
```python
avg_bid_ask = mean(Ask_Price - Bid_Price) / mean(Mid_Price)
```
- **Datos disponibles**: Columnas 'Bid Price' y 'Ask Price'
- **Interpretación**: Coste implícito de transacción

### **Métrica 3: Volumen Negociado (Información Adicional Deseable)**
- **Outstanding Amount como proxy**: Emisiones más grandes suelen ser más líquidas
- **Edad de la emisión**: Bonos más antiguos pueden ser menos líquidos
- **Número de market makers**: Más dealers = mayor liquidez

### **Métrica 4: Concentración por Emisión**
```python
concentration_risk = max(portfolio['weight'])
```
- **Lógica**: Posiciones grandes son más difíciles de liquidar

### **OTRAS FUENTES DE INFORMACIÓN QUE SERÍAN ÚTILES:**
1. **Volumen diario negociado** por ISIN
2. **Número de trades diarios**
3. **Horario de negociación** (algunas plataformas)
4. **Presencia en índices principales** (mayor liquidez)
5. **Ratings de liquidez** de agencias especializadas
6. **Spread histórico bid-ask** y su volatilidad

##### <font color=#CC6600>Backtest</font>

In [ ]:
## 🔄 **METODOLOGÍA DE BACKTESTING - PASO A PASO:**

### **PASO 1: Definir Fechas de Rebalanceo**
```python
rebalance_dates = pd.date_range(start='2023-10-01', end='2025-09-30', freq='M')
```
- **Frecuencia**: Mensual (último día de cada mes)
- **Período**: 2 años de historia disponible

### **PASO 2: Por Cada Fecha de Rebalanceo**
```python
for date in rebalance_dates:
    # 2.1 Filtrar universo disponible en esa fecha
    available_bonds = get_bonds_with_prices(date)
    
    # 2.2 Recalcular métricas (yield, duración, PD, etc.)
    update_bond_metrics(available_bonds, date)
    
    # 2.3 Aplicar TODAS las restricciones
    eligible_bonds = apply_constraints(available_bonds)
    
    # 2.4 Optimizar pesos (maximizar yield/duración ratio)
    weights = optimize_portfolio(eligible_bonds)
    
    # 2.5 Calcular retorno del período
    period_return = calculate_returns(weights, date_prev, date)
```

### **PASO 3: Verificación Continua de Restricciones**
- **Durante el mes**: Monitorear si alguna restricción se viola
- **Ejemplo**: Si un bono es downgraded y viola límite HY
- **Acción**: Rebalanceo extraordinario si es necesario

### **PASO 4: Cálculo de Performance**
```python
# Retorno total
nav_series = (1 + monthly_returns).cumprod()

# Métricas de riesgo
volatility = monthly_returns.std() * sqrt(12)
sharpe_ratio = (annual_return - risk_free) / volatility
max_drawdown = calculate_max_drawdown(nav_series)
```

### **PASO 5: Comparación vs Benchmark**
- **Benchmark**: RECMTREU Index (Total Return)
- **Métricas**: Tracking error, information ratio, correlación
- **Análisis**: Períodos de outperformance/underperformance

### **CONSIDERACIONES IMPORTANTES:**
1. **Costes de transacción**: Bid-ask spread en cada rebalanceo
2. **Look-ahead bias**: Solo usar información disponible en cada fecha
3. **Survivorship bias**: Incluir bonos que defaultearon o maduraron
4. **Slippage**: Impacto de mercado en posiciones grandes

### <font color=#336699>7. Cobertura tipos de interés</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Utiliza alguno de los siguientes instrumentos de los que te hemos dado para cubrir la duración (sensibilidad de tipos de interés) de la cartera que has construido según el mandato. Asume una inversión en la cartera de 10 millones:

- Futuros sobre el *Schatz* (ticker: DU1) - Duración a 01/10/2025: 1.92
- Futuros sobre el *BOBL* (ticker: OE1) - Duración a 01/10/2025: 5.44
- Futuros sobre el *BUND* (ticker: RX1) - Duración a 01/10/2025: 10

*Contract size* en todos los casos: 100,000 euros

Investiga sobre estos instrumentos antes de tomar la decisión. Razona tu elección del instrumento y el número de contratos que has decidido comprar/vender.

- ¿Qué pasaría si comprásemos/vendiésemos 100 futuros?
- ¿Se te ocurre algún otro instrumento con el que cubrir la sensibilidad a los tipos de interés de la cartera?

</div>

In [6]:
# EJERCICIO 7: COBERTURA DE TIPOS DE INTERÉS
# Implementación usando módulo profesional

from hedging_analysis import InterestRateHedging, run_hedging_analysis

# Ejecutar análisis completo de cobertura
hedging_analyzer = run_hedging_analysis()

✅ Datos cargados: 514 observaciones
📈 Datos históricos cargados:
  DU1 Comdty: 509 obs., rango 104.82 - 107.67
  OE1 Comdty: 509 obs., rango 115.21 - 120.44
  RX1 Comdty: 509 obs., rango 127.07 - 138.64
🎯 EJERCICIO 7: COBERTURA DE TIPOS DE INTERÉS

📊 DATOS DE LA CARTERA A CUBRIR:
Valor de la cartera: 10,000,000 EUR
Duración de la cartera: 0.95 años
Riesgo de duración: 9,500,000 EUR·años
🎯 CÁLCULO DE HEDGE RATIOS:
Fórmula: Hedge Ratio = (Portfolio Duration × Portfolio Value) / (Future Duration × Contract Size)

Schatz (DU1):
  Contratos necesarios: 49.5 contratos
  Exposición: 9,500,000 EUR·años
  Eficiencia: 100.0%

BOBL (OE1):
  Contratos necesarios: 17.5 contratos
  Exposición: 9,500,000 EUR·años
  Eficiencia: 100.0%

BUND (RX1):
  Contratos necesarios: 9.5 contratos
  Exposición: 9,500,000 EUR·años
  Eficiencia: 100.0%

📋 ANÁLISIS COMPARATIVO:

Schatz (DU1) ⭐⭐⭐⭐⭐:
  ✅ Ventajas:
    • Duración más cercana a cartera (1.92 vs 0.95)
    • Máxima liquidez
    • Menor basis risk
    • Mej

In [ ]:
# Análisis específicos usando el módulo

# 1. Obtener recomendación detallada
recommendation = hedging_analyzer.get_recommendation()
print("🎯 RECOMENDACIÓN DETALLADA:")
print(f"Instrumento: {recommendation['instrument']}")
print(f"Contratos: {recommendation['direction']} {recommendation['contracts']} contratos")
print(f"Eficiencia de cobertura: {recommendation['efficiency']:.1f}%")

# 2. Análisis de escenarios personalizados
print("\n📊 ANÁLISIS DE 50 CONTRATOS ÓPTIMOS:")
optimal_results = hedging_analyzer.scenario_analysis(50)

# 3. Acceso a instrumentos alternativos
alternatives = hedging_analyzer.get_alternative_instruments()
print("\n🛠️ NÚMERO DE ALTERNATIVAS DISPONIBLES:", len(alternatives))

🎯 CÁLCULO DE HEDGE RATIOS:
Fórmula: Hedge Ratio = (Portfolio Duration × Portfolio Value) / (Future Duration × Contract Size)

Schatz (DU1):
  Contratos necesarios: 49.5 contratos
  Exposición: 9,500,000 EUR·años
  Eficiencia: 100.0%

BOBL (OE1):
  Contratos necesarios: 17.5 contratos
  Exposición: 9,500,000 EUR·años
  Eficiencia: 100.0%

BUND (RX1):
  Contratos necesarios: 9.5 contratos
  Exposición: 9,500,000 EUR·años
  Eficiencia: 100.0%

🎯 RECOMENDACIÓN DETALLADA:
Instrumento: Schatz (DU1)
Contratos: VENDER 49 contratos
Eficiencia de cobertura: 99.0%

📊 ANÁLISIS DE 50 CONTRATOS ÓPTIMOS:

📊 ANÁLISIS DE ESCENARIO: 50 CONTRATOS
📈 Schatz (DU1):
  Exposición total: 9,600,000 EUR·años
  Ratio vs cartera: 1.0x
  P&L por cambio de yield:
    -1.0%: +96,000 EUR
    -0.5%: +48,000 EUR
    +0.0%: -0 EUR
    +0.5%: -48,000 EUR
    +1.0%: -96,000 EUR

📈 BOBL (OE1):
  Exposición total: 27,200,000 EUR·años
  Ratio vs cartera: 2.9x
  P&L por cambio de yield:
    -1.0%: +272,000 EUR
    -0.5%: +136,

## 🎯 **RESPUESTA TEÓRICA COMPLETA - EJERCICIO 7**

### **📊 DECISIÓN DE COBERTURA**

**Instrumento elegido: SCHATZ (DU1)**
- **Contratos a vender**: 50 contratos (redondeado desde 49.5)
- **Exposición**: 50 × 100,000 × 1.92 = 9,600,000 EUR·años
- **Eficiencia**: 101% de cobertura (9.6M vs 9.5M EUR·años de la cartera)

### **📋 JUSTIFICACIÓN DE LA ELECCIÓN**

#### **1. Matching de Duración Óptimo**
- Cartera: 0.95 años vs Schatz: 1.92 años
- Ratio 2:1 es manejable y predecible
- BOBL (5.44) y BUND (10.0) crean exceso de cobertura

#### **2. Minimización del Basis Risk** 
- Correlación alta entre bonos corporativos EUR y Bunds alemanes
- Curva alemana como referencia para toda la eurozona
- Menor distorsión por efectos de crédito específicos

#### **3. Eficiencia Operativa**
- Máxima liquidez en Eurex
- Spreads bid-ask más estrechos
- Facilidad para ajustes dinámicos

### **⚖️ ANÁLISIS DEL ESCENARIO 100 CONTRATOS**

Si compráramos/vendiéramos 100 contratos:

| Instrumento | Exposición | Ratio vs Cartera | Riesgo |
|-------------|------------|------------------|--------|
| **Schatz** | 19.2M EUR·años | 2.0x | Sobrecubierto moderado |
| **BOBL** | 54.4M EUR·años | 5.7x | Muy peligroso |  
| **BUND** | 100.0M EUR·años | 10.5x | Extremadamente peligroso |

**Conclusión**: 100 contratos fijos generarían más riesgo que protección.

### **🛠️ INSTRUMENTOS ALTERNATIVOS RECOMENDADOS**

1. **Interest Rate Swaps (IRS)**: Para carteras >50M EUR
2. **ETFs de Bonos**: Para inversores retail sin acceso a derivados
3. **Opciones sobre Futuros**: Para cobertura asimétrica
4. **Futuros de otros países**: Para diversificación geográfica

### **📈 IMPLEMENTACIÓN PRÁCTICA**

```
Operación: VENDER 50 contratos Schatz (DU1)
Objetivo: Neutralizar sensibilidad a tipos de interés
Monitoreo: Rebalancear mensualmente según duración de cartera
Stop-loss: Si basis risk > 50bps durante 5 días consecutivos
```

### **🎯 RESULTADO ESPERADO**

- **Cobertura**: 101% de la sensibilidad a tipos
- **Coste**: Spread bid-ask + comisiones ≈ 5-10 bps
- **Beneficio**: Cartera neutralizada ante movimientos de tipos
- **Riesgo residual**: Basis risk corporativo vs soberano ≈ 10-20 bps

### <font color=#336699>8. Cobertura crédito</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Utiliza alguno de los siguientes instrumentos de los que te hemos dado para cubrir el riesgo de crédito de la cartera que has construido según el mandato. Asume una inversión en la cartera de 10 millones:

- ITRAXX Main (ticker: ITRX EUR CDSI GEN 5Y Corp)
- ITRAXX XOVER (ticker: ITRX XOVER CDSI GEN 5Y Corp)

Estos índices cotizan en forma de spread, en puntos básicos. La sensibilidad del valor del swap (CDS) la vamos a asumir en 4,500€ al punto básico asumiendo una inversión de 10 millones.

Investiga sobre estos instrumentos antes de tomar la decisión. Razona tu elección del instrumento y el nominal que has decidido comprar/vender.

- ¿Tiene sentido plantear esta cobertura total?
- ¿Con qué otros instrumentos podrías cubrir el riesgo de crédito?

</div>

In [ ]:
Investigación y elección del instrumento

ITRAXX Main (ITRX EUR CDSI GEN 5Y Corp) es un índice CDS que referencia empresas europeas de grado de inversión (IG). Es muy líquido y ampliamente utilizado para cubrir el riesgo de crédito IG.
ITRAXX XOVER (ITRX XOVER CDSI GEN 5Y Corp) referencia empresas europeas de alto rendimiento (HY), adecuado para cubrir el riesgo de crédito HY.
Ambos índices cotizan en forma de spread (puntos básicos), y la sensibilidad del valor del swap para una inversión de 10 millones de euros se asume en 4.500 € por punto básico.
Razonamiento para la elección y el nominal

Para cubrir la cartera, se igualan los pesos IG y HY con el índice correspondiente:
Comprar protección en ITRAXX Main por la parte IG (por ejemplo, si IG = 80%, nominal = 8.000.000 €).
Comprar protección en ITRAXX XOVER por la parte HY (por ejemplo, si HY = 20%, nominal = 2.000.000 €).
Este enfoque alinea la cobertura con el perfil crediticio de la cartera, usando instrumentos líquidos y estandarizados.

¿Tiene sentido plantear una cobertura total?

Una cobertura total elimina todo el riesgo de crédito, protegiendo ante ampliaciones de spread pero también eliminando el potencial de mejora crediticia.
Puede justificarse si el mandato exige control estricto del riesgo o si las condiciones de mercado sugieren riesgo elevado.
Normalmente, es preferible una cobertura parcial o dinámica, equilibrando reducción de riesgo y potencial de rentabilidad.

¿Con qué otros instrumentos podrías cubrir el riesgo de crédito?

CDS de nombre individual sobre las mayores exposiciones.
Opciones de crédito (si existen).
Índices CDS sectoriales.
Diversificación de la cartera por sector/país.
Reducción directa de la exposición HY.

### <font color=#336699>9. Estrategia propia</font>

<style>.gray {background-color: #595959}

</style><div class="gray">

❕💬 Plantea tu propia estrategia con la información que tienes. Puede ser una estrategia direccional, de valor relativo, que hayas visto o no en clase; pero siempre razonando tu planteamiento.

</div>

In [ ]:
Estrategia: Valor Relativo IG/HY + Barbell de Duración
1. Valor Relativo IG/HY:
Sobrepondero bonos de grado de inversión (IG) y limito la exposición a alto rendimiento (HY) al máximo permitido (10%). Esto permite aprovechar la mayor seguridad y liquidez de los IG, mientras se obtiene algo de rentabilidad adicional con HY, pero sin asumir riesgos excesivos.

2. Estructura Barbell de Duración:
Divido la cartera entre bonos muy cortos (1-2 años) y bonos cercanos al límite de duración (3 años). Así, optimizo el carry y la convexidad, haciendo la cartera menos sensible a subidas de tipos, pero capaz de aprovechar movimientos de mercado.

3. Diversificación:
Mantengo exposición a distintos sectores y países para reducir riesgos específicos y mejorar la resiliencia de la cartera.

4. Coberturas y Simulación:
Utilizo futuros y CDS índices para ajustar la sensibilidad a tipos y crédito según expectativas. Simulo escenarios de subida de tipos y ampliación de spreads para analizar el impacto en la cartera.

5. Cumplimiento del Mandato:
La estrategia respeta todas las restricciones: máximo 20 bonos, duración ≤ 3 años, HY ≤ 10%, y diversificación.

Razonamiento:
Esta estrategia busca equilibrio entre prudencia y rentabilidad, adaptándose al entorno de tipos bajos y riesgo de subidas, y a spreads ajustados. Permite flexibilidad y protección ante shocks de mercado, y puede ajustarse dinámicamente según evolucionen las condiciones.